src/OpenConext/Profile/Value/Arp.php

```diff
@@ -0,0 +1,147 @@
+    {
+        $arpCollection = [];
+
+        // Create Attribute instances for all attributes
+        foreach ($arp as $attributeName => $attributeDefinitionInformation) {
+            $attributeSource = 'idp';
+            $attributeDefinition = new AttributeDefinition($attributeName, $attributeName, $attributeName);
+
+            // When the dictionary is available. Lookup the attribute in dictionary to load friendly attribute names
+            if (!is_null($dictionary)) {
+                try {
+                    $attributeDefinition = $dictionary->getAttributeDefinitionByUrn($attributeName);
+                } catch (UnknownUrnException $exception) {
+                    // Use the previously created attributeDefinition.
+                }
+            }
+
+            if (isset($attributeDefinitionInformation[0]['source'])) {
+                $attributeSource = $attributeDefinitionInformation[0]['source'];
+            }
+
+            // The arp is grouped on attribute source
+            $arpCollection[$attributeSource][] = new Attribute($attributeDefinition, $attributeDefinitionInformation);
+        }
+
+        $this->arp = $arpCollection;
+    }
+
+    /**
+     * Tests the structure of the Arp attribute information.
+     *
+     * This information should be an array, should have the attribute names as its keys and have array values.
+     * These array values can have two keys (value & source) the values aof these entries should be string.
+     *
+     * Example of a valid attribute information array:
+     *
+     * [
+     *   'urn.mace.email' => [
+     *     ['value' => '*'],
+     *   ],
+     *   'urn.mace.eduPersonTargetedId' => [
+     *     [
+     *       'value' => '*',
+     *       'source' => 'sab',
+     *     ]
+     *   ],
+     *   'urn.mace.orcid' => [
+     *     [
+     *       'value' => 'orcid.org/\d{4}-\d{4}',
+     *       'source' => 'orcid',
+     *     ],
+     *     [
+     *       'value' => 'sandbox.orcid.org/\d{4}-\d{4}',
+     *       'source' => 'orcid',
+     *     ],
+     *   ],
+     * ]
+     *
+     * @param array $attributeInformation
+     * @return bool
+     */
+    private static function isValidAttribute(array $attributeInformation)
+    {
+        $validKeys = ['value', 'source'];
+
+        foreach ($attributeInformation as $attributeInformationEntry) {
+            foreach ($attributeInformationEntry as $key => $attributeConfigValue) {
+                if (!in_array($key, $validKeys)) {
+                    return false;
+                }
+                if (!is_string($attributeConfigValue)) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    /**
+     * @return array
+     */
+    public function getNonIdpAttributes()
+    {
+        $attributes = $this->getAttributesGroupedBySource();
+        if (isset($attributes['idp'])) {
```

Comment:
This is redundant: just unset(..) will do :)

Reply:
TIL